modules/navier_stokes/src/fvkernels/INSFVTKESourceSink.C

```diff
@@ -0,0 +1,267 @@
+      auto distance = _dist[_current_elem][i];
+
+      auto y_plus = NS::findyPlus(_mu(makeElemArg(_current_elem), state),
+                                  _rho(makeElemArg(_current_elem), state),
+                                  std::max(parallel_speed, 1e-10),
+                                  distance);
+
+      y_plus_vec.push_back(y_plus);
+
+      ADReal velocity_grad_norm = parallel_speed / _dist[_current_elem][i];
+
+      /// Do not erase!!
+      /// More complete expansion for velocity gradient. Leave commented for now.
+      //// Will be useful later when doing two-phase or compressible flow
+      // ADReal velocity_grad_norm_sq =
+      //     Utility::pow<2>(_u_var->gradient(makeElemArg(_current_elem), state) *
+      //                     _normal[_current_elem][i]);
+
+      // if (_dim >= 2)
+      //   velocity_grad_norm_sq +=
+      //       Utility::pow<2>(_v_var->gradient(makeElemArg(_current_elem), state) *
+      //                       _normal[_current_elem][i]);
+
+      // if (_dim >= 3)
+      //   velocity_grad_norm_sq +=
+      //       Utility::pow<2>(_w_var->gradient(makeElemArg(_current_elem), state) *
+      //                       _normal[_current_elem][i]);
+
+      // ADReal velocity_grad_norm = std::sqrt(velocity_grad_norm_sq);
+
+      velocity_grad_norm_vec.push_back(velocity_grad_norm);
+
+      tot_weight += 1.0;
+    }
+
+    for (unsigned int i = 0; i < y_plus_vec.size(); i++)
+    {
+      auto y_plus = y_plus_vec[i];
+
+      auto fi = _face_infos[_current_elem][i];
+      const bool defined_on_elem_side = _var.hasFaceSide(*fi, true);
+      const Elem * const loc_elem = defined_on_elem_side ? &fi->elem() : fi->neighborPtr();
+      Moose::FaceArg facearg = {
+          fi, Moose::FV::LimiterType::CentralDifference, false, false, loc_elem};
+      ADReal wall_mut = _mu_t(facearg, state);
+
+      auto destruction_visc =
+          2.0 * wall_mut / Utility::pow<2>(_dist[_current_elem][i]) / tot_weight;
+      auto destruction_log = std::pow(_C_mu, 0.75) * _rho(makeElemArg(_current_elem), state) *
+                             std::pow(_var(makeElemArg(_current_elem), old_state), 0.5) /
+                             (_von_karman * _dist[_current_elem][i]) / tot_weight;
+
+      if (y_plus < 11.25)
+        destruction += destruction_visc;
+      else
+        destruction += destruction_log;
+
+      production += wall_mut * velocity_grad_norm_vec[i] * std::pow(_C_mu, 0.25) /
+                    std::sqrt(_var(makeElemArg(_current_elem), old_state) + 1e-10) /
+                    (_von_karman * _dist[_current_elem][i]) / tot_weight;
+    }
+
+    residual = (destruction - production) * _var(makeElemArg(_current_elem), state);
+  }
+  else
+  {
+
+    const auto & grad_u = _u_var->adGradSln(_current_elem, state);
+    auto Sij_xx = 2.0 * grad_u(0);
+    ADReal Sij_xy = 0.0;
+    ADReal Sij_xz = 0.0;
+    ADReal Sij_yy = 0.0;
+    ADReal Sij_yz = 0.0;
+    ADReal Sij_zz = 0.0;
```

Comment:
Same comment as before, would be much nicer if you used ADTensorValue here. You have used these in the MomentumDiffusion complete expansion I think.

Reply:
I tried and it is slowing down runtime. Will revisit later in the future.